backend/src/main/scala/bloop/reporter/Reporter.scala

```diff
@@ -30,6 +30,8 @@ abstract class Reporter(
     val logger: Logger,
     override val cwd: AbsolutePath,
     override val config: ReporterConfig,
+    // FIXME: _problems needs to be thread-safe. Shall we use a Java ConcurrentLinkedQueue or would you prefer something else?
+    //        Keeping the mutable buffer can cause Hydra to crash because of off-by-one exceptions thrown for instance by Buffer#clear
     val _problems: mutable.Buffer[ProblemPerPhase] = mutable.ArrayBuffer.empty
```

Comment:
Hmmm, would it be possible to parameterize the reporter with a buffer interface of sorts so that we can switch the implementation to `ConcurrentLinkedQueue` when using Hydra?

Reply:
Sounds good. I'm going to have an abstract Buffer interface with two implementations: one using the current Scala `mutable.Buffer` and another one using a `ConcurrentLinkedQueue`.

Reply:
Yes, I think having a simple interface with only the methods we use should be enough, thanks Mirco 👍

Reply:
Yep, that's exactly what I have. The implementations will be value classes, so we should have no allocation penalty (not that I think it really matters here, but I thought you'd still like that :)).

Reply:
Great, another way we might want to implement this is by avoiding creating an interface for a buffer and just having some abstract methods to add/manage problems in the reporter interface that are implemented in a `SingleThreadedReporter` and a `ConcurrentReporter`. I almost find this solution better than creating an interface for buffers. What do you think?

Reply:
I've just pushed the changes that implement the solution we initially discussed. TBH, I'd actually suggest to make the concurrent implementation the default, as I don't believe users will experience any performance penalty, and it would keep the code simpler (as there is only one implementation and all `Reporter` implementations simply need to be thread-safe)

Reply:
> I'd actually suggest to make the concurrent implementation the default
  
I prefer to err on the side of caution and given that we don't have the time to run benchmarks to make sure this is the case that we keep these two implementations separate

Reply:
Sure. I'm totally ok with keeping things as they are. Shall we mark this conversation resolved?